File: backend/src/main/java/com/decmoe47/todo/service/impl/TodoListServiceImpl.java
package com.decmoe47.todo.service.impl;

import cn.hutool.core.bean.BeanUtil;
import com.decmoe47.todo.constant.enums.ErrorCodeEnum;
import com.decmoe47.todo.exception.ErrorResponseException;
import com.decmoe47.todo.model.dto.TodoListAddDTO;
import com.decmoe47.todo.model.dto.TodoListDeleteDTO;
import com.decmoe47.todo.model.dto.TodoListUpdateDTO;
import com.decmoe47.todo.model.entity.Todo;
import com.decmoe47.todo.model.entity.TodoList;
import com.decmoe47.todo.model.entity.User;
import com.decmoe47.todo.model.vo.TodoListVO;
import com.decmoe47.todo.repository.TodoListRepository;
import com.decmoe47.todo.repository.TodoRepository;
import com.decmoe47.todo.repository.UserRepository;
import com.decmoe47.todo.service.TodoListService;
import com.decmoe47.todo.util.SecurityUtil;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.stereotype.Service;

import java.util.List;

@RequiredArgsConstructor
@Service
public class TodoListServiceImpl implements TodoListService {

    private final TodoListRepository todoListRepo;
    private final TodoRepository todoRepo;
    private final UserRepository userRepo;
    private final CacheManager cacheManager;

    @Override
    public List<TodoListVO> getCustomTodoLists() {
        long userId = SecurityUtil.getCurrentUserId();
        List<TodoList> lists = todoListRepo.findByCreatedByIdAndInboxFalse(userId);
        return BeanUtil.copyToList(lists, TodoListVO.class);
    }

    @Override
    public TodoListVO addTodoList(TodoListAddDTO todoListAddDTO) {
        User user = userRepo.findById(SecurityUtil.getCurrentUserId())
                .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.USER_NOT_FOUND));
        TodoList todoList = BeanUtil.toBean(todoListAddDTO, TodoList.class);
        todoList.setCreatedBy(user);
        todoList = todoListRepo.save(todoList);
        return BeanUtil.toBean(todoList, TodoListVO.class);
    }

    @Override
    public TodoListVO updateTodoList(TodoListUpdateDTO todoListUpdateDTO) {
        TodoList todoList = todoListRepo.findById(todoListUpdateDTO.getId())
                .orElseThrow(() -> new ErrorResponseException(ErrorCodeEnum.TODO_LIST_NOT_FOUND));
        BeanUtil.copyProperties(todoListUpdateDTO, todoList);
        todoList = todoListRepo.save(todoList);
        return BeanUtil.toBean(todoList, TodoListVO.class);
    }

    @Transactional
    @Override
    public void deleteTodoList(TodoListDeleteDTO todoListDeleteDTO) {
        long userId = SecurityUtil.getCurrentUserId();
        List<Todo> todos = todoRepo.findByCreatedByIdAndBelongedList_Id(userId, todoListDeleteDTO.getId());    // TODO: 改用mybatis plus select单列
        List<Long> todoIds = todos.stream().map(Todo::getId).toList();

        todoRepo.deleteByBelongedList_Id(todoListDeleteDTO.getId());
        todoListRepo.deleteById(todoListDeleteDTO.getId());
        clearCache(todoListDeleteDTO.getId(), todoIds);
    }

    private void clearCache(String todoListId, List<Long> todoIds) {
        Cache todoListAccess = cacheManager.getCache("todoListAccess");
        if (todoListAccess != null) {
            todoListAccess.evict(SecurityUtil.getCurrentUserId() + "_" + todoListId);
        }

        Cache todoAccess = cacheManager.getCache("todoAccess");
        if (todoAccess != null) {
            for (Long todoId : todoIds) {
                todoAccess.evict(SecurityUtil.getCurrentUserId() + "_" + todoId);
            }
        }
    }
}

File: backend/src/main/kotlin/com/decmoe47/todo/service/impl/TodoListServiceImpl.kt
package com.decmoe47.todo.service.impl

import com.decmoe47.todo.annotation.ReadOnlyTransactionalService
import com.decmoe47.todo.constant.enums.ErrorCode
import com.decmoe47.todo.exception.ErrorResponseException
import com.decmoe47.todo.model.entity.AuditableEntity
import com.decmoe47.todo.model.entity.TodoList
import com.decmoe47.todo.model.mapper.toTodoListResponse
import com.decmoe47.todo.model.request.TodoListAddRequest
import com.decmoe47.todo.model.request.TodoListDeleteRequest
import com.decmoe47.todo.model.request.TodoListUpdateRequest
import com.decmoe47.todo.model.response.TodoListResponse
import com.decmoe47.todo.repository.TodoListRepository
import com.decmoe47.todo.repository.TodoRepository
import com.decmoe47.todo.repository.UserRepository
import com.decmoe47.todo.service.TodoListService
import com.decmoe47.todo.util.SecurityUtil
import org.springframework.cache.CacheManager
import org.springframework.transaction.annotation.Transactional

@ReadOnlyTransactionalService
class TodoListServiceImpl(
    private val todoListRepo: TodoListRepository,
    private val todoRepo: TodoRepository,
    private val userRepo: UserRepository,
    private val cacheManager: CacheManager,
) : TodoListService {
    override fun getAll(): List<TodoListResponse> {
        val userId = SecurityUtil.getCurrentUserId()
        val lists = todoListRepo.selectAll(userId)
        return lists.map { it.toTodoListResponse() }
    }

    @Transactional
    override fun add(request: TodoListAddRequest): TodoListResponse {
        val userId = SecurityUtil.getCurrentUserId()
        val user = userRepo.first(userId) ?: throw ErrorResponseException(ErrorCode.USER_NOT_FOUND)

        val todoList = TodoList(
            name = request.name,
            auditable = AuditableEntity(createdBy = user.id),
        )
        return todoListRepo.save(todoList).toTodoListResponse()
    }

    @Transactional
    override fun update(request: TodoListUpdateRequest): TodoListResponse {
        val todoList = todoListRepo.first(request.id)
            ?: throw ErrorResponseException(ErrorCode.TODO_LIST_NOT_FOUND)
        val updated = todoList.copy(
            name = request.name,
            auditable = todoList.auditable.copy(updatedBy = SecurityUtil.getCurrentUserId()),
        )
        return todoListRepo.update(updated).toTodoListResponse()
    }

    @Transactional
    override fun delete(request: TodoListDeleteRequest) {
        val todos = todoRepo.select(request.id)
        val todoIds = todos.map { it.id }

        todoRepo.deleteByBelongedListId(request.id)
        todoListRepo.delete(request.id)
        clearCache(request.id, todoIds)
    }

    private fun clearCache(todoListId: Long, todoIds: List<Long>) {
        val userId = SecurityUtil.getCurrentUserId()
        val todoListAccess = cacheManager.getCache("todoListAccess")
        todoListAccess?.evict("${userId}_${todoListId}")

        val todoAccess = cacheManager.getCache("todoAccess")
        if (todoAccess != null) {
            todoIds.forEach { todoId ->
                todoAccess.evict("${userId}_${todoId}")
            }
        }
    }
}
